### DISCUSSION TITLE
WCHAR32 breaks WGPU (and expectedly so)

### DISCUSSION THREAD

**Discussion by andreamancuso** [ocornut/imgui#7649] [New Users Build/Link/Run issues ONLY!]
Hi,

I apologise in advance if this is the wrong place/topic. I started learning WebAssembly, C++, Dear Imgui about a month ago - please bear with me.

My setup:

- CMake
- vcpkg
- target: emscripten

I am currently [using imgui as a vcpkg package](https://vcpkg.io/en/package/imgui). Currently the following features are enabled: `glfw-binding`, `opengl3-binding`, `wgpu-binding`. Today I tried enabling `wchar32`. Everything builds fine.

Unfortunately, as I run the application I get a WebGPU issue: "exceeded max texture size". I believe this is expected.

The question I have is: will `wchar32` support ever be available while using WebGPU as renderer?

Thank you in advance for your time.

**Comment by ocornut** [comment]
I don't know exactly what you mean by "I tried enabling `wchar32`" but this seems like an issue specific to however vcpkg handle those flags.

I see no reason for `IMGUI_USE_WCHAR32` compile-time option to have any side-effect on imgui_impl_wgpu backend. Those things are completely unrelated. There's no issue of ImWchar in that backend.

**Reply by andreamancuso** [reply]
Apologies. You are of course correct.

I was trying to embed [Material Design Icons](https://github.com/juliettef/IconFontCppHeaders?tab=readme-ov-file#pictogrammers-material-design-icons--mdi-) but I since reverted back to using font-awesome icons and now enabling `wchar32` works like a charm.

**Reply by ocornut** [reply]
It seems like your issue isn't related to wchar32 but to the size of your texture.
Since you are not providing code we can only guess what's happening, but here are a few hints:
https://github.com/ocornut/imgui/blob/master/docs/FONTS.md#4-font-atlas-texture-fails-to-upload-to-gpu

Eventually we'll fix this font system :)

**Reply by andreamancuso** [reply]
I feel quite dumb for not having realized what is actually happening... Anyway, should you ever wish to look at it, here's the code: https://github.com/andreamancuso/react-wasm/pull/29

I am very happy with font awesome icons. That said, it would be amazing to be able to use Material Design icons in the future.

Thank you again for your time, best wishes.

**Reply by ocornut** [reply]
> That said, it would be amazing to be able to use Material Design icons in the future.

You can use it today. It's not about the font per-se but about amount of glyphs and glyphes sizes. 
As mentioned in FAQ entry, using font_cfg->OversampleH = 1 may help.

You can also set atlas->TexDesiredWidth to max width supported by device.
We currently don't obtain this info from backends, probably we should when we finish work on #3761

**Reply by andreamancuso** [reply]
You're right. `font_cfg->OversampleH = 1` does fix it. I do see 100% CPU utilisation for about 15 seconds before the application starts getting rendered, however.
I'll try the `atlas->TexDesiredWidth` change.

**Reply by ocornut** [reply]
> I do see 100% CPU utilisation for about 15 seconds before the application starts getting rendered, however.

That's bad :(
The headers suggest it has ~2240 glyphes vs ~1400 for FontAwesome6 so I imagine it is due to the kind of per-glyph details.
Our current system or prebaking everything would makes this redhibitory until we start baking glyph on demand, sorry.

**Reply by andreamancuso** [reply]
Truth be told, who needs 2240 fonts in a single application ... ?

In 2008, where bundle size of web apps truly mattered, we used to generate our own version of glyph-based web fonts. A bit cumbersome but highly effective.

At any rate, thank you again for looking into this!
